src/network_service/sending_targets_cache.rs

```diff
@@ -0,0 +1,137 @@
+    cache: HashMap<u64, Vec<(ConnectionInfo, TargetState)>>,
+}
+
+impl SendingTargetsCache {
+    pub fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+        }
+    }
+
+    pub fn insert_message(
+        &mut self,
+        msg_id: u64,
+        initial_targets: Vec<ConnectionInfo>,
+        dg_size: usize,
+    ) {
+        let targets = initial_targets
+            .into_iter()
+            .enumerate()
+            .map(|(idx, tgt_info)| {
+                (
+                    tgt_info,
+                    if idx < dg_size {
+                        TargetState::Sending(0)
+                    } else {
+                        TargetState::Failed(0)
+                    },
+                )
+            })
+            .collect();
+        let _ = self.cache.insert(msg_id, targets);
+    }
+
+    fn target_states<'a>(
+        &'a self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &'a (ConnectionInfo, TargetState)> {
+        self.cache.get(&msg_id).into_iter().flatten()
+    }
+
+    fn target_states_mut<'a>(
+        &'a mut self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &'a mut (ConnectionInfo, TargetState)> {
+        self.cache.get_mut(&msg_id).into_iter().flatten()
+    }
+
+    fn fail_target(&mut self, msg_id: u64, target: SocketAddr) {
+        let _ = self
+            .target_states_mut(msg_id)
+            .find(|(info, _state)| info.peer_addr() == target)
+            .map(|(_info, state)| match *state {
+                TargetState::Failed(_) => {
+                    log_or_panic!(LogLevel::Error, "Got a failure from a failed target!");
+                }
+                TargetState::Sending(x) => {
+                    *state = TargetState::Failed(x + 1);
+                }
+                TargetState::Sent => {
+                    log_or_panic!(
+                        LogLevel::Error,
+                        "A target that should no longer fail - failed!"
+                    );
+                }
+            });
+    }
+
+    fn get_next_target(&self, msg_id: u64) -> Option<ConnectionInfo> {
+        self.target_states(msg_id)
+            .filter(|(_info, state)| !state.is_complete())
+            .filter_map(|(info, state)| match state {
+                TargetState::Failed(x) => Some((info, *x)),
+                _ => None,
+            })
+            .min_by_key(|&(_info, num)| num)
+            .map(|(info, _num)| info)
+            .cloned()
+    }
+
+    fn should_drop(&self, msg_id: u64) -> bool {
+        self.target_states(msg_id)
+            .all(|(_info, state)| state.is_complete())
+    }
+
+    pub fn target_failed(&mut self, msg_id: u64, target: SocketAddr) -> Option<ConnectionInfo> {
+        self.fail_target(msg_id, target);
+        if self.should_drop(msg_id) {
+            let _ = self.cache.remove(&msg_id);
+        }
+        // if we dropped the msg_id above, this would have returned None even if we hadn't
+        self.get_next_target(msg_id)
```

Comment:
We are not setting that target to Sending, I assume that will cause issues.

Reply:
:man_facepalming: You're right. Let me try with this change.